blocks/datasets/__init__.py

```diff
@@ -0,0 +1,233 @@
+            e.g. the number of examples to return, or the indices of a
+            particular minimbatch of examples.
+        sources : tuple of strings, optional
+            The data sources to return. By default all data sources are
+            returned.
+
+        .. todo::
+
+           A way for the dataset to communicate which kind of requests it
+           accepts, and a way to communicate what kind of request is being
+           sent when supporting multiple.
+
+        Returns
+        -------
+        tuple
+            A tuple of data.
+
+        """
+        raise NotImplementedError
+
+
+class DataStream(object):
+    """A stream of data.
+
+    A data stream implements the ``__iter__`` protocol, returning a new
+    iterator over the data each time it gets called (in the case of finite
+    data). A single pass over one of these iterators represents an *epoch*.
+
+    The data stream also opens a given dataset and maintains the state
+    (e.g. file handles) if necessary.
+
+    A data stream can provide an iterator over a dataset, but it can also
+    wrap another datastream e.g. to provide caching or some sort of
+    pre-processing.
+
+    Parameters
+    ----------
+    data : :class:`Dataset` or :class:`DataStream`
+        An object that implements the :meth:`open` and :meth:`get_data`
+        methods. Usually this is a :class:`Dataset` instance, but when
+        chaining iterators to e.g. perform caching this can also be a
+        :class:`DataStream` object.
+    iteration_scheme : :class:`IterationScheme`, optional
+        The iteration scheme to use when retrieving data. Note that not all
+        datasets support the same iteration schemes, some data sets require
+        one, and others don't support any.
+    sources : tuple of strings, optional
+        The sources of data to return. By default, all sources of the
+        dataset are requested.
+
+    """
+    def __init__(self, data, iteration_scheme=None, sources=None):
+        update_instance(self, locals())
+        self.data_state = None
+
+    def __iter__(self):
+        self.data_state = self.data.open(self.data_state)
+        # TODO Allow for multiple child datasets/data streams, so that they
+        # can be combined
+        if isinstance(self.data, DataStream):
+            self.child_iterator = iter(self.data)
+        return DataIterator(self, iter(self.iteration_scheme)
+                            if self.iteration_scheme else None, self.sources)
+
+    @property
+    def sources(self):
+        if getattr(self, '_sources', None) is None:
+            return self.data.sources
+
+    @sources.setter
+    def sources(self, sources):
+        self._sources = sources
+
+    def open(self, state=None):
+        return self.data.open(state)
+
+    def close(self, satte):
+        self.data.close(self.data_state)
+
+    def get_data(self, state=None, request=None, sources=None):
```

Comment:
Have a safe trip!

I can not think of any usecase for this agnosticity. Looks like data stream can be of two types: those that initiate data flow from a dataset and those that transform already existing streams. E.g. your `CachedDataStream` in practice can only wrap another data stream as it uses a `child_iterator`.

Based on these thoughts I propose the following changes:

* make `get_data` use `self.data_state` and fetch the next request from the request iterator. The only parameter left would then be `sources`.

*  the `DataIterator.__next__` then becomes that simple:
```python
def __next__(self):
     return self.data_stream.get_data(self.sources)
```

* this new `get_data` method can be used instead of `next(self.child_iterator)` by transforming data streams like your `CachedDataStream`